### ISSUE TITLE
Loading custom DLL outside Tabular Editor 3 folder?

### ISSUE THREAD

**Issue by bernatagulloesbrina** [TabularEditor/TabularEditor3#340]
Hi, I am now testing the custom dll thing in my corporate PC where I do not have permissons to modify the folder of Tabular Editor 3 under program files and indeed I cannot manage to make it work.  

I downloaded the portable version, and copied the dll there. Closing and opening did the trick and managed to make it work on the portable version. 

I guess that Tabular Editor cannot bypass the admin restrictions, so it's not possible to create a "import dll" function or similar. So maybe a special folder can be used instead? What about the folder where the macros and all is stored? 

some solution on this front would help spread the use of custom dlls and more sofisticated scripts

**Comment by bernatagulloesbrina** [comment]
by the way, this is the error 

> Script Execution Failed
> 
> Could not load file or assembly 'ExtendedTOMWrapper, Version=22.1.4.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. El sistema no puede encontrar el archivo especificado.

so maybe there's a way to build the dll so that all dependencies are found? 
I have tried referencing a couple more dlls from Tabular Editor 3, but did not cause any difference.

**Comment by otykier** [comment]
I assume you already tried executing the script with syntax such as:

```csharp
#r "C:\Users\DanielOtykier\Documents\TestCustomDll.dll"

using TestCustomDll;

string x;
StringUtils.InitString("a","b","c", out x, "d");
```

If this doesn't work on the corporate PC, then I'm afraid there's not much we can do, as it boils down to whether the .NET framework allows dynamic loading of arbitrary DLLs. At least when the DLL is located in the same folder as Tabular Editor, or when the DLL is registered in the Global Assembly Cache, you can be sure that it was put there by someone with administrative permissions. Otherwise, the DLL could have been put there by anyone with access to the machine, and so the .NET framework does not deem it "safe" to load.

In fact, this is the reason why we generally don't encourage (and document) the use of custom DLL's in scripts, since there are these restrictions to be aware about (not to mention issues with versioning of nested dependencies, etc.).

We do have plans to provide better scripting facilities within TE3 itself, for example by allowing you to create and save scripts that contain custom namespaces, classes, types, etc., which you can then call from other scripts. Unfortunately, we can't provide a timeline yet, since we have other backlog items with higher priority.

**Comment by bernatagulloesbrina** [comment]
Hi Daniel, 

yes, I tried with the full path and did not work. 
we've been with Xavi trying all the ideas he had (and he had quite a few) but didn't pull it off. 

We even tried modifying the TabularEditor3.exe.config file and add (following[ this .net documentation)](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.loadfrom?redirectedfrom=MSDN&view=net-6.0#remarks) 
`<loadFromRemoteSources enabled="true"/>  `
but did not work either -- he was quite puzzled by that. 

we managed to load the references programatically inside the script by pointing to the path, but then cannot be really used 

at this point, is there a way to add all the class code inside the script ? my guess so far is that no, it's not possible to define inline classes right? 

if I continue the develpment for users who can manage to get the dll in the tabular editor folder, will it be possible to port to the future enhanced scripting so that everyone can run it? 

I'm glad it's on the roadmap, now trying to decide the best approach in the meantime 

Thank you for your patience (with me) and your efforts with the product!

**Comment by otykier** [comment]
Since writing my reply above, one of our devs may have found a fix that allows loading of DLLs from arbitrary locations. Needs some more testing, but if it works, we'll include it in the next update (both TE2 and TE3) and let you know.

**Comment by bernatagulloesbrina** [comment]
Awesome!

**Comment by otykier** [comment]
Update [3.2.0](https://github.com/TabularEditor/TabularEditor3/releases/tag/3.2.0) now allows you to load assemblies from any location by specifying the full path of the DLL.

**Comment by bernatagulloesbrina** [comment]
great great great! Kudos to the tabular editor team!

**Comment by otykier** [comment]
Just FYI a similar update was just released for TE2: [2.16.6](https://github.com/TabularEditor/TabularEditor/releases/tag/2.16.6).
